transport/tchannel/handler.go

```diff
@@ -0,0 +1,181 @@
+//
+// We need to do this so that we can change the return type of call.Response()
+// to match inboundCall's Response().
+type tchannelCall struct{ *tchannel.InboundCall }
+
+func (c tchannelCall) Response() inboundCallResponse {
+	return c.InboundCall.Response()
+}
+
+// handler wraps a transport.Handler into a TChannel Handler.
+type handler struct {
+	Handler transport.Handler
+}
+
+func (h handler) Handle(ctx context.Context, call *tchannel.InboundCall) {
+	h.handle(ctx, tchannelCall{call})
+}
+
+func (h handler) handle(ctx context.Context, call inboundCall) {
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		call.Response().SendSystemError(tchannel.ErrTimeoutRequired)
+		return
+	}
+
+	headers, err := readHeaders(call.Format(), call.Arg2Reader)
+	if err != nil {
+		call.Response().SendSystemError(tchannel.NewSystemError(
+			tchannel.ErrCodeUnexpected, "failed to read headers: %v", err))
+		return
+	}
+
+	body, err := call.Arg3Reader()
+	if err != nil {
+		call.Response().SendSystemError(tchannel.NewSystemError(
+			tchannel.ErrCodeUnexpected, "failed to read body: %v", err))
+		return
+	}
+	defer body.Close()
+
+	rw := newResponseWriter(call)
+	defer rw.Close() // TODO(abg): log if this errors
+
+	treq := &transport.Request{
+		Caller:    call.CallerName(),
+		Service:   call.ServiceName(),
+		Encoding:  transport.Encoding(call.Format()),
+		Procedure: call.MethodString(),
+		Headers:   headers,
+		Body:      body,
+		TTL:       deadline.Sub(time.Now()),
+	}
+
+	if err := h.Handler.Handle(ctx, treq, rw); err != nil {
+		call.Response().SendSystemError(tchannel.NewSystemError(
+			tchannel.ErrCodeUnexpected, "internal error: %v", err))
+		return
+	}
+}
+
+type responseWriter struct {
+	failedWith   error
+	bodyWriter   tchannel.ArgWriter
+	format       tchannel.Format
+	headers      transport.Headers
+	response     inboundCallResponse
+	wroteHeaders bool
+}
+
+func newResponseWriter(call inboundCall) *responseWriter {
+	return &responseWriter{
+		response: call.Response(),
+		headers:  make(transport.Headers),
+		format:   call.Format(),
+	}
+}
+
+func (rw *responseWriter) AddHeaders(h transport.Headers) {
+	if rw.wroteHeaders {
+		panic("AddHeaders() cannot be called after calling Write().")
+	}
+	for k, v := range h {
+		rw.headers.Set(k, v)
+	}
+}
+
+func (rw *responseWriter) Write(s []byte) (int, error) {
+	if rw.failedWith != nil {
+		return 0, rw.failedWith
+	}
+
+	if !rw.wroteHeaders {
+		rw.wroteHeaders = true
+		if err := writeHeaders(rw.format, rw.headers, rw.response.Arg2Writer); err != nil {
+			// TODO(abg): wrap this and the Arg3Writer error in something.
+			rw.failedWith = err
+			return 0, err
+		}
+	}
+
+	if rw.bodyWriter == nil {
+		var err error
+		rw.bodyWriter, err = rw.response.Arg3Writer()
+		if err != nil {
+			rw.failedWith = err
+			return 0, err
+		}
+	}
+
+	return rw.bodyWriter.Write(s)
+}
+
+func (rw *responseWriter) Close() error {
+	if rw.bodyWriter != nil {
```

Comment:
Good call. We'll call close but return the `Write()` error if it had failed earlier.